File: coauthor/latex/texcount.py
import os

from ..logger import logger

from ..utils.exec import execute_command


def get_texcount(filePaths: list[str] | str, merge: bool = False) -> str | None:
    """Run texcount on LaTeX files and return combined statistics output, optionally merging included files."""
    if not isinstance(filePaths, list):
        filePaths = [filePaths]

    all_outputs = []
    for filePath in filePaths:
        if not os.path.exists(filePath):
            logger.warning(f"Warning: File {filePath} does not exist.")
            continue

        if ".tex" not in filePath:
            logger.warning(f"Error: File {filePath} is not a LaTeX file. Skipping.")
            continue

        command = ["texcount"]
        if merge:
            command.append("-merge")
        command.append(filePath)

        success, stdout, stderr = execute_command(command, capture_output=True)
        if success:
            all_outputs.append(f"Tex Count Results for {filePath}:\n{stdout}")
        else:
            logger.error(f"Error getting tex count for {filePath}")
            logger.error(f"Stdout: {stdout}")
            logger.error(f"Stderr: {stderr}")

    if all_outputs:
        combined_output = "\n\n".join(all_outputs)
        logger.info(f"Combined Tex Count Results:\n{combined_output}")
        return combined_output
    return None


def get_texcountStats(inputFiles: str | list[str]) -> str | None:
    """Run texcount on LaTeX files and return formatted statistics with XML-style tags."""
    if isinstance(inputFiles, str):
        inputFiles = [inputFiles]

    texcountStats = get_texcount(inputFiles)
    return f"Tex Count Statistics:<texcount>\n{texcountStats}\n</texcount>\n\n" if texcountStats else None

File: coauthor-for-vs-code/src/latex/texcount.ts
// Local imports - log
import * as logger from '../logger/logUtils';

// Local imports - utilities
import { fileExists } from '../utils/fileUtils';
import { executeCommand } from '../utils/execUtils';
import { checkToolInstalled } from './texTools';

const CHANNEL = 'LaTeXCommands';
logger.initialize(CHANNEL);

/**
 * Get full statistics for LaTeX documents using the texcount Perl script
 * @param filePaths Single file path or array of file paths
 * @param merge Whether to merge included files in the count
 * @param channel The channel to use for logging
 * @returns Promise<string | null> String containing full texcount output for all files, or null if an error occurred
 */
export async function getTexCount(
  filePaths: string | string[],
  merge: boolean = false,
  channel: string = CHANNEL,
): Promise<string | null> {
  try {
    // Check if texcount is installed
    if (!(await checkToolInstalled('texcount'))) {
      return null;
    }

    // Convert single path to array
    const paths = Array.isArray(filePaths) ? filePaths : [filePaths];
    const allOutputs: string[] = [];

    for (const filePath of paths) {
      if (!(await fileExists(filePath))) {
        logger.warn(channel, `Warning: File ${filePath} does not exist.`);
        continue;
      }

      if (!filePath.endsWith('.tex')) {
        logger.warn(
          channel,
          `Error: File ${filePath} is not a LaTeX file. Skipping.`,
        );
        continue;
      }

      const command = ['texcount'];
      if (merge) {
        command.push('-merge');
      }
      command.push(`"${filePath}"`);

      const result = await executeCommand(command, {
        channel,
        truncate: false, // Don't truncate texcount output as we need the full statistics
      });
      if (result.success && result.stdout) {
        allOutputs.push(`Tex Count Results for ${filePath}:\n${result.stdout}`);
        logger.debug(channel, `Successfully counted ${filePath}`);
      } else {
        logger.error(channel, `Error getting tex count for ${filePath}`);
        if (result.stdout) logger.error(channel, `Stdout: ${result.stdout}`);
        if (result.stderr) logger.error(channel, `Stderr: ${result.stderr}`);
      }
    }

    if (allOutputs.length > 0) {
      const combinedOutput = allOutputs.join('\n\n');
      logger.info(channel, `Combined Tex Count Results:\n${combinedOutput}`);
      return combinedOutput;
    }

    return null;
  } catch (err) {
    logger.error(
      channel,
      `Error in getTexCount: ${err instanceof Error ? err.message : String(err)}`,
    );
    return null;
  }
}

/**
 * Run texcount on LaTeX files and return formatted statistics with XML-style tags
 * @param filePaths Single file path or array of file paths
 * @param channel The channel to use for logging
 * @returns Promise<string | null> String containing formatted texcount statistics with XML tags, or null if an error occurred
 */
export async function getTexCountStats(
  filePaths: string | string[],
  channel: string = CHANNEL,
): Promise<string | null> {
  const texcountStats = await getTexCount(filePaths, false, channel);
  return texcountStats
    ? `Tex Count Statistics:<texcount>\n${texcountStats}\n</texcount>\n\n`
    : null;
}
